File: tests/tooling/perf/sort.c
/* The C twin of sort.fpp: the same median-of-middle quicksort, same keys. */
#include <stdio.h>
#define N 2000000
#define REPS 6
static int a[N];

static void qsort_(int lo, int hi) {
    if (lo < hi) {
        int mid = lo + (hi - lo) / 2;
        int p = a[mid];
        int i = lo, j = hi;
        while (i <= j) {
            while (a[i] < p) i++;
            while (a[j] > p) j--;
            if (i <= j) { int t = a[i]; a[i] = a[j]; a[j] = t; i++; j--; }
        }
        if (lo < j) qsort_(lo, j);
        if (i < hi) qsort_(i, hi);
    }
}

int main(void) {
    long long acc = 0;
    for (int rep = 0; rep < REPS; rep++) {
        unsigned s = 2463534242u;
        for (int i = 0; i < N; i++) {
            s ^= s << 13; s ^= s >> 17; s ^= s << 5;
            a[i] = (int)(s % 10000000u);
        }
        qsort_(0, N - 1);
        acc += a[0] + a[N / 2] + a[N - 1];
    }
    printf("%.0f\n", (double)acc);
    return 0;
}

File: tests/tooling/perf/sort.rs
// The Rust twin of sort.fpp: in-place quicksort over an i32 slice, the same
// xorshift keys. Bounds checks are ON, including in the partition scans that
// no analysis can prove — the same two F++ still checks.
const N: usize = 2000000;
const REPS: usize = 6;

fn qsort(a: &mut [i32], lo: isize, hi: isize) {
    if lo < hi {
        let mid = lo + (hi - lo) / 2;
        let p = a[mid as usize];
        let mut i = lo;
        let mut j = hi;
        while i <= j {
            while a[i as usize] < p { i += 1; }
            while a[j as usize] > p { j -= 1; }
            if i <= j {
                a.swap(i as usize, j as usize);
                i += 1;
                j -= 1;
            }
        }
        if lo < j { qsort(a, lo, j); }
        if i < hi { qsort(a, i, hi); }
    }
}

fn main() {
    let mut a = vec![0i32; N];
    let mut acc: i64 = 0;
    for _ in 0..REPS {
        let mut s: u32 = 2463534242;
        for i in 0..N {
            s ^= s << 13; s ^= s >> 17; s ^= s << 5;
            a[i] = (s % 10000000) as i32;
        }
        qsort(&mut a, 0, (N - 1) as isize);
        acc += a[0] as i64 + a[N / 2] as i64 + a[N - 1] as i64;
    }
    println!("{:.0}", acc as f64);
}
